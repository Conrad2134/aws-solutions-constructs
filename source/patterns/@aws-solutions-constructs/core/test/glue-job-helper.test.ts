/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import { ResourcePart, SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { CfnJob, CfnJobProps } from '@aws-cdk/aws-glue';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { Stack } from "@aws-cdk/core";
import * as defaults from '..';

// --------------------------------------------------------------
// Test minimal deployment with no properties
// --------------------------------------------------------------
test('Test minimal deployment with no properties', () => {
  // Stack
  const stack = new Stack();
  const _jobID = 'glueetl';

  const _jobRole = new Role(stack, 'CustomETLJobRole', {
    assumedBy: new ServicePrincipal('glue.amazonaws.com')
  });

  const cfnJobProps: CfnJobProps = defaults.DefaultGlueJobProps(_jobRole.roleArn, {
    name: _jobID,
    pythonVersion: '3',
    scriptLocation: 's3://fakelocation/script'
  }, 'testETLJob');

  defaults.buildGlueJob(stack, { glueJobProps: cfnJobProps });
  // Assertion 1
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  // Assertion 2
  expect(stack).toHaveResourceLike('AWS::Glue::Job', {
    Type: "AWS::Glue::Job",
    Properties: {
      Command: {
        Name: "glueetl",
        PythonVersion: "3",
        ScriptLocation: "s3://fakelocation/script"
      },
      Role: {
        "Fn::GetAtt": [
          "CustomETLJobRole90A83A66",
          "Arn"
        ]
      },
      SecurityConfiguration: "testETLJob"
    }
  }, ResourcePart.CompleteDefinition);
});

// --------------------------------------------------------------
// Create a Glue job outside the construct, should not create a new one
// --------------------------------------------------------------
test('Create a Glue Job outside the construct', () => {
  // Stack
  const stack = new Stack();
  const _existingCfnJob = new CfnJob(stack, 'ExistingJob', {
    command: {
        name: 'pythonshell',
        pythonVersion: '2',
        scriptLocation: 's3://existingFakeLocation/existingScript'
    },
    role: new Role(stack, 'ExistingJobRole', {
        assumedBy: new ServicePrincipal('glue.amazon.com'),
        description: 'Existing role'
    }).roleArn,
    glueVersion: '1',
    allocatedCapacity: 2,
    maxCapacity: 4,
    numberOfWorkers: 2,
    workerType: 'Standard'
  });

  defaults.buildGlueJob(stack, {
      existingCfnJob: _existingCfnJob
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(stack).toHaveResourceLike('AWS::Glue::Job', {
    Type: "AWS::Glue::Job",
    Properties: {
      AllocatedCapacity: 2,
      Command: {
        Name: "pythonshell",
        PythonVersion: "2",
        ScriptLocation: "s3://existingFakeLocation/existingScript",
      },
      GlueVersion: "1",
      MaxCapacity: 4,
      NumberOfWorkers: 2,
      Role: {
        "Fn::GetAtt": [
            "ExistingJobRole8F750976",
            "Arn",
        ],
      },
      WorkerType: "Standard",
    }
  }, ResourcePart.CompleteDefinition);
});

// --------------------------------------------------------------
// Provide additional parameters other than default ones
// --------------------------------------------------------------
test('Test custom deployment properties', () => {
  // Stack
  const stack = new Stack();
  const _commandName = 'glueetl';

  const cfnJobProps: CfnJobProps = {
    command: {
      name: _commandName,
      pythonVersion: '3',
      scriptLocation: 's3://existingFakeLocation/existingScript'
    },
    role: new Role(stack, 'ExistingJobRole', {
      assumedBy: new ServicePrincipal('glue.amazon.com'),
      description: 'Existing role'
    }).roleArn,
    glueVersion: '1',
    allocatedCapacity: 2,
    maxCapacity: 4,
    numberOfWorkers: 2,
    workerType: 'Standard'
  };

  defaults.buildGlueJob(stack, { glueJobProps: cfnJobProps });
  // Assertion 1
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();

  // check if Glue Job Resource was created correctly
  expect(stack).toHaveResourceLike('AWS::Glue::Job', {
    Properties: {
      AllocatedCapacity: 2,
      Command: {
        Name: "glueetl",
        PythonVersion: "3",
        ScriptLocation: "s3://existingFakeLocation/existingScript",
      },
      GlueVersion: "1",
      MaxCapacity: 4,
      NumberOfWorkers: 2,
      Role: {
        "Fn::GetAtt": [
          "ExistingJobRole8F750976",
          "Arn",
        ],
      },
      SecurityConfiguration: "ETLJobSecurityConfig",
      WorkerType: "Standard",
    },
    Type: "AWS::Glue::Job"
  }, ResourcePart.CompleteDefinition);

  // check if the role is created
  expect(stack).toHaveResourceLike('AWS::IAM::Role',         {
    Type: "AWS::IAM::Role",
    Properties: {
      AssumeRolePolicyDocument: {
        Statement: [
          {
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
                Service: "glue.amazon.com"
            }
          }
        ],
        Version: "2012-10-17"
      },
      Description: "Existing role"
    }
  }, ResourcePart.CompleteDefinition);

  // check if the security config is created
  expect(stack).toHaveResourceLike('AWS::Glue::SecurityConfiguration', {
    Properties: {
      EncryptionConfiguration: {
        JobBookmarksEncryption: {
          JobBookmarksEncryptionMode: "CSE-KMS",
          KmsKeyArn: {
            "Fn::Join": [
              "", [
                "arn:", {
                    Ref: "AWS::Partition",
                },
                ":kms:", {
                    Ref: "AWS::Region",
                },
                ":", {
                    Ref: "AWS::AccountId",
                },
                ":alias/aws/glue",
              ],
            ],
          },
        },
        S3Encryptions: [{
            S3EncryptionMode: "SSE-S3",
        }],
      },
      Name: "ETLJobSecurityConfig",
    },
    Type: "AWS::Glue::SecurityConfiguration",
  }, ResourcePart.CompleteDefinition);
});

test('Do no supply glueJobProps or existingCfnJob and error out', () => {
  const stack = new Stack();
  try {
    defaults.buildGlueJob(stack, {});
  } catch(error) {
    expect(error).toBeInstanceOf(Error);
  }
});