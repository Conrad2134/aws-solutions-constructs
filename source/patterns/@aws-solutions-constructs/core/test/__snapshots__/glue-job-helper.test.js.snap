// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Database 1`] = `
Object {
  "Resources": Object {
    "GlueDatabase": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseInput": Object {
          "Description": "A Glue database table",
        },
      },
      "Type": "AWS::Glue::Database",
    },
  },
}
`;

exports[`Create a Glue Job outside the construct 1`] = `
Object {
  "Resources": Object {
    "ExistingJob": Object {
      "Properties": Object {
        "AllocatedCapacity": 2,
        "Command": Object {
          "Name": "pythonshell",
          "PythonVersion": "2",
          "ScriptLocation": "s3://existingFakeLocation/existingScript",
        },
        "GlueVersion": "1",
        "MaxCapacity": 4,
        "NumberOfWorkers": 2,
        "Role": Object {
          "Fn::GetAtt": Array [
            "ExistingJobRole8F750976",
            "Arn",
          ],
        },
        "WorkerType": "Standard",
      },
      "Type": "AWS::Glue::Job",
    },
    "ExistingJobRole8F750976": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazon.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Existing role",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`Create table 1`] = `
Object {
  "Resources": Object {
    "GlueDatabase": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseInput": Object {
          "Description": "A Glue database table",
        },
      },
      "Type": "AWS::Glue::Database",
    },
    "GlueTable": Object {
      "Properties": Object {
        "CatalogId": Object {
          "Ref": "AWS::AccountId",
        },
        "DatabaseName": Object {
          "Ref": "GlueDatabase",
        },
        "TableInput": Object {
          "Parameters": Object {
            "classication": "json",
          },
          "StorageDescriptor": Object {
            "Columns": Array [
              Object {
                "Comment": "",
                "Name": "id",
                "Type": "int",
              },
              Object {
                "Comment": "",
                "Name": "name",
                "Type": "string",
              },
              Object {
                "Comment": "",
                "Name": "address",
                "Type": "string",
              },
              Object {
                "Comment": "",
                "Name": "value",
                "Type": "int",
              },
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
            "Location": "testStream",
            "NumberOfBuckets": -1,
            "OutputFormat": "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
            "Parameters": Object {
              "endpointUrl": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "https://kinesis.",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ".amazonaws.com",
                  ],
                ],
              },
              "streamName": "testStream",
              "typeOfData": "kinesis",
            },
            "SerdeInfo": Object {
              "Parameters": Object {
                "paths": "id,name,address,value",
              },
              "SerializationLibrary": "org.openx.data.jsonserde.JsonSerDe",
            },
          },
          "TableType": "EXTERNAL_TABLE",
        },
      },
      "Type": "AWS::Glue::Table",
    },
  },
}
`;

exports[`Provide script location directory location 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersd683bbd12df11a4cc2038d55bd98ef859fb10f79b53aeb89145ac615a6671174ArtifactHashB7D5B24E": Object {
      "Description": "Artifact hash for asset \\"d683bbd12df11a4cc2038d55bd98ef859fb10f79b53aeb89145ac615a6671174\\"",
      "Type": "String",
    },
    "AssetParametersd683bbd12df11a4cc2038d55bd98ef859fb10f79b53aeb89145ac615a6671174S3Bucket770047E3": Object {
      "Description": "S3 bucket for asset \\"d683bbd12df11a4cc2038d55bd98ef859fb10f79b53aeb89145ac615a6671174\\"",
      "Type": "String",
    },
    "AssetParametersd683bbd12df11a4cc2038d55bd98ef859fb10f79b53aeb89145ac615a6671174S3VersionKey9115CE5E": Object {
      "Description": "S3 key for asset version \\"d683bbd12df11a4cc2038d55bd98ef859fb10f79b53aeb89145ac615a6671174\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "JobRole014917C6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Service role that Glue custom ETL jobs will assume for exeuction",
      },
      "Type": "AWS::IAM::Role",
    },
    "JobRoleDefaultPolicy5DE0D8F9": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParametersd683bbd12df11a4cc2038d55bd98ef859fb10f79b53aeb89145ac615a6671174S3Bucket770047E3",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParametersd683bbd12df11a4cc2038d55bd98ef859fb10f79b53aeb89145ac615a6671174S3Bucket770047E3",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "JobRoleDefaultPolicy5DE0D8F9",
        "Roles": Array [
          Object {
            "Ref": "JobRole014917C6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`Test GlueJobCommandProperty creation 1`] = `
Object {
  "Resources": Object {
    "JobRole014917C6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Service role that Glue custom ETL jobs will assume for exeuction",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`Test custom deployment properties 1`] = `
Object {
  "Resources": Object {
    "ETLJob": Object {
      "Properties": Object {
        "AllocatedCapacity": 2,
        "Command": Object {
          "Name": "glueetl",
          "PythonVersion": "3",
          "ScriptLocation": "s3://existingFakeLocation/existingScript",
        },
        "GlueVersion": "1",
        "MaxCapacity": 4,
        "NumberOfWorkers": 2,
        "Role": Object {
          "Fn::GetAtt": Array [
            "ExistingJobRole8F750976",
            "Arn",
          ],
        },
        "SecurityConfiguration": "ETLJobSecurityConfig",
        "WorkerType": "Standard",
      },
      "Type": "AWS::Glue::Job",
    },
    "ExistingJobRole8F750976": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazon.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Existing role",
      },
      "Type": "AWS::IAM::Role",
    },
    "GlueSecurityConfig": Object {
      "Properties": Object {
        "EncryptionConfiguration": Object {
          "JobBookmarksEncryption": Object {
            "JobBookmarksEncryptionMode": "CSE-KMS",
            "KmsKeyArn": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":kms:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":alias/aws/glue",
                ],
              ],
            },
          },
          "S3Encryptions": Array [
            Object {
              "S3EncryptionMode": "SSE-S3",
            },
          ],
        },
        "Name": "ETLJobSecurityConfig",
      },
      "Type": "AWS::Glue::SecurityConfiguration",
    },
    "LogPolicy9292E033": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:/aws-glue/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogPolicy9292E033",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "ExistingJobRole8F750976",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`Test deployment with role creation 1`] = `
Object {
  "Resources": Object {
    "CustomETLJobRole90A83A66": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ETLJob": Object {
      "Properties": Object {
        "Command": Object {
          "Name": "glueetl",
          "PythonVersion": "3",
          "ScriptLocation": "s3://fakelocation/script",
        },
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomETLJobRole90A83A66",
            "Arn",
          ],
        },
        "SecurityConfiguration": "testETLJob",
      },
      "Type": "AWS::Glue::Job",
    },
    "LogPolicy9292E033": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:/aws-glue/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogPolicy9292E033",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "CustomETLJobRole90A83A66",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
